File: PracticaMonitoresenJava/Buffer.java
public class Buffer {
    private final int[] buffer;
    private int count, in, out;
    private final int SIZE;
    
    public Buffer(int size) {
        SIZE = size;
        buffer = new int[SIZE];
        count = in = out = 0;
    }

    // El método synchronized asegura que solo un hilo pueda producir a la vez
    public synchronized void producir(int item) throws InterruptedException {
        while (count == SIZE) {
            // Esperar si el buffer está lleno
            System.out.println("Buffer lleno. Productor esperando...");
            wait();
        }
        
        // Producir: insertar el item en el buffer
        buffer[in] = item;
        in = (in + 1) % SIZE;
        count++;
        
        System.out.println("Producido: " + item);
        
        // Notificar al consumidor que puede consumir
        notify();
    }

    // El método synchronized asegura que solo un hilo pueda consumir a la vez
    public synchronized int consumir() throws InterruptedException {
        while (count == 0) {
            // Esperar si el buffer está vacío
            System.out.println("Buffer vacío. Consumidor esperando...");
            wait();
        }
        
        // Consumir: retirar el item del buffer
        int item = buffer[out];
        out = (out + 1) % SIZE;
        count--;
        
        System.out.println("Consumido: " + item);
        
        // Notificar al productor que puede producir
        notify();
        
        return item;
    }
}

File: PracticaMonitoresenJava/Buffer.py
import threading
import time
from queue import Queue

class Buffer:
    def __init__(self, size):
        self.size = size
        self.buffer = Queue(maxsize=size)
        self.lock = threading.Lock()
        self.condition = threading.Condition(self.lock)

    def producir(self, item):
        with self.condition:
            while self.buffer.full():
                print("Buffer lleno. Productor esperando...")
                self.condition.wait()
            
            self.buffer.put(item)
            print(f"Producido: {item}")
            
            # Notificar al consumidor que puede consumir
            self.condition.notify()

    def consumir(self):
        with self.condition:
            while self.buffer.empty():
                print("Buffer vacío. Consumidor esperando...")
                self.condition.wait()
            
            item = self.buffer.get()
            print(f"Consumido: {item}")
            
            # Notificar al productor que puede producir
            self.condition.notify()
            
            return item

def productor(buffer):
    for i in range(10):
        buffer.producir(i)
        time.sleep(1)

def consumidor(buffer):
    for _ in range(10):
        buffer.consumir()
        time.sleep(2)

def main():
    buffer = Buffer(5)

    # Crear y ejecutar los hilos productor y consumidor
    hilo_productor = threading.Thread(target=productor, args=(buffer,))
    hilo_consumidor = threading.Thread(target=consumidor, args=(buffer,))

    hilo_productor.start()
    hilo_consumidor.start()

    # Esperar a que ambos hilos terminen
    hilo_productor.join()
    hilo_consumidor.join()

if __name__ == "__main__":
    main()
